### PR TITLE
user:group + permissions + homedir - make composer download work for /usr/local/bin and custom user:group who cannot RW to target_dir

### PR DESCRIPTION
**Pull request by rednut**
Had some glitches when trying to deposit composer into a directory not RW as intended user, also issues with odd files (like `cache` and `.htaccess` being dumped in `target_dir`) plus permissions when HOME not set 

- add ability to define GROUP aswell as USER permissions

- fix permission denied when  `target_dir` is not writable by `composer::user` / `composer::group` - wget composer into target_dir as root, then change permissions using `File`

- fix permissions should use puppet native `File` type to apply permission and owner/group to target file

- if `composer::user` does not have a home dir set, the `COMPOSER_HOME` was depositing a `cache` and `.htaccess` in `target_dir` - fixed by setting the `COMPOSER_HOME` to /tmp for the `selfupdate` call

### CODE REVIEW

**Comment by deviantintegral** [comment]
I've run into a similar issue where $HOME is /home/vagrant, and not /root, so provisioned boxes have a .composer directory that isn't writable by the vagrant user. I'm working around it by removing .composer entirely after provisioning since I'm shipping a basebox, but this isn't a great solution for updates supplied via provisioning.

File: README.md [hunk 1]
```diff
@@ -29,6 +29,19 @@ instance. So, please, run the following command:
...
+---------------
+
+Example configuration using Hiradata style config:
+
+```
+composer::user: 'vagrant'
+composer::group: 'vagrant'
+composer::target_dir: '/usr/local/bin/'
+composer::auto_update: true
+```
```
Comment (Ma27) [inline README.md]: when writing docs I'd like to see how to configure it using pure puppet and not with hiera. Hiera is a nice tool of course, but not everyone uses it.

File: .travis.yml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 language: ruby
-
+sudo: required
```
Comment (Ma27) [inline .travis.yml]: is this really necessary?

File: manifests/init.pp [hunk 3]
```diff
@@ -29,11 +32,14 @@
   $target_dir   = 'UNDEF',
   $command_name = 'UNDEF',
   $user         = 'UNDEF',
+  $group        = 'UNDEF',
```
Comment (Ma27) [inline manifests/init.pp]: please use for new params the undef constant. In a new major release I'll fix the legacy params to use the undef parameter, too

File: manifests/init.pp [hunk 4]
```diff
@@ -49,28 +55,45 @@
     default => $user
   }
 
+  $composer_group = $group ? {
+    'UNDEF' => $::composer::params::group,
+    default => $group
+  }
+
+  # common dependency
+  package { 'patch':
```
Comment (Ma27) [inline manifests/init.pp]: for what is this needed?

File: manifests/init.pp [hunk 5]
```diff
@@ -49,28 +55,45 @@
...
+    ensure => latest
+  }
+
+
+  # fetch composer into target_dir
   wget::fetch { 'composer-install':
     source      => $::composer::params::phar_location,
     destination => "${composer_target_dir}/${composer_command_name}",
-    execuser    => $composer_user,
+    execuser    => 'root',
```
Comment (Ma27) [inline manifests/init.pp]: :-1: if you'd like to separate the downloading user from *$composer_user* then please create a new parameter

File: manifests/init.pp [hunk 6]
```diff
@@ -49,28 +55,45 @@
...
 
-  exec { 'composer-fix-permissions':
-    command => "chmod a+x ${composer_command_name}",
-    path    => '/usr/bin:/bin:/usr/sbin:/sbin',
-    cwd     => $composer_target_dir,
-    user    => $composer_user,
-    unless  => "test -x ${composer_target_dir}/${composer_command_name}",
+
+  # apply user and group permissions to downloaded composer using native puppet File type
+  file { 'composer-fix-permissions':
```
Comment (Ma27) [inline manifests/init.pp]: why is a file type needed instead of exec?

File: manifests/init.pp [hunk 7]
```diff
@@ -49,28 +55,45 @@
...
   }
 
-  exec { 'composer-fix-permissions':
-    command => "chmod a+x ${composer_command_name}",
-    path    => '/usr/bin:/bin:/usr/sbin:/sbin',
-    cwd     => $composer_target_dir,
-    user    => $composer_user,
-    unless  => "test -x ${composer_target_dir}/${composer_command_name}",
+
+  # apply user and group permissions to downloaded composer using native puppet File type
```
Comment (Ma27) [inline manifests/init.pp]: if you leave comments, then please don't tell us WHAT you're doing, but WHY you're doing this

File: manifests/init.pp [hunk 8]
```diff
@@ -49,28 +55,45 @@
...
+    recurse => false,
     require => Wget::Fetch['composer-install'],
   }
 
+  # run self update when requested
   if $auto_update {
     exec { 'composer-update':
-      command     => "${composer_command_name} self-update",
-      environment => [ "COMPOSER_HOME=${composer_target_dir}" ],
+      command     => "php -d allow_url_fopen=1 ${composer_target_dir}/${composer_command_name} self-update",
```
Comment (Ma27) [inline manifests/init.pp]: does this mean that it is not possible to access the composer by its command name??
 If it is possible then you should keep the old code. If not then I'm :-1:

**Comment by Ma27** [comment]
a rebase is also necessary.

**Comment by Ma27** [comment]
ping @rednut

**Comment by Ma27** [comment]
closing due to no response by the PR author.

**Pull request closed without merging**